custom_components/rct_power/__init__.py

```diff
@@ -7,39 +7,49 @@
 
 from __future__ import annotations
 
-import asyncio
 import logging
+from dataclasses import dataclass
 from datetime import timedelta
 from typing import Literal
 
 from homeassistant.config_entries import ConfigEntry
 from homeassistant.core import HomeAssistant
 from homeassistant.exceptions import ConfigEntryNotReady
 from homeassistant.helpers.typing import ConfigType
+from homeassistant.util.hass_dict import HassEntryKey
 
 from .lib.api import RctPowerApiClient
-from .lib.const import DOMAIN, PLATFORMS, STARTUP_MESSAGE
-from .lib.context import RctPowerContext
-from .lib.domain_data import get_domain_data
+from .lib.const import DOMAIN, PLATFORMS, STARTUP_MESSAGE, EntityUpdatePriority
 from .lib.entities import all_entity_descriptions
-from .lib.entity import EntityUpdatePriority, resolve_object_infos
+from .lib.entity import resolve_object_infos
 from .lib.entry import RctPowerConfigEntryData, RctPowerConfigEntryOptions
 from .lib.update_coordinator import RctPowerDataUpdateCoordinator
 
 SCAN_INTERVAL = timedelta(seconds=30)
+RCT_DATA_KEY: HassEntryKey[RctData] = HassEntryKey(DOMAIN)
 
 _LOGGER: logging.Logger = logging.getLogger(__package__)
 
+type RctConfigEntry = ConfigEntry[RctData]
+
+
+@dataclass
+class RctData:
+    update_coordinators: dict[EntityUpdatePriority, RctPowerDataUpdateCoordinator]
+
 
 async def async_setup(hass: HomeAssistant, config: ConfigType):
     """Set up this integration using YAML is not supported."""
     return True
 
 
-async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> Literal[True]:
+async def async_setup_entry(
+    hass: HomeAssistant, entry: RctConfigEntry
+) -> Literal[True]:
     """Set up this integration using UI."""
-    if len(domain_data := get_domain_data(hass)) == 0:
+    if not (data := hass.data.setdefault(DOMAIN, {})):
         _LOGGER.info(STARTUP_MESSAGE)
+        data["startup_message"] = True
```

Comment:
Any thoughts on not logging that message at all? Is there a real benefit that warrants the log file spam?

Reply:
Thought about that as well. I don't see any real benefit in keeping the log message. Home Assistant itself will log a warning for custom integrations already
> [homeassistant.loader] We found a custom integration rct_power which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant

and if users go to the integrations page, `Documentation` and `Known issues` will link to this repo as well.

I've already planned a small logger improvement, let's remove it then.

Reply:
Agreed - do you want to do this here or in a follow-up?

Reply:
In a follow-up.

Reply:
> In a follow-up.

Opened #455 for it. _It depends on #454 to be merged first in order to avoid conflicts._